### PR TITLE
TOF: add trigger efficiency

### PR DESCRIPTION
**Pull request by ercolessi**
@njacazio

### CODE REVIEW

File: Modules/TOF/include/TOF/TaskRaw.h [hunk 1]
```diff
@@ -83,6 +83,7 @@ class RawDataDecoder final : public DecoderBase
   Counter<nequipments, nullptr> mCounterNoisyChannels;    /// Counter for noisy channels
   Counter<1024, nullptr> mCounterTimeBC;                  /// Counter for the Bunch Crossing Time
   Counter<91, nullptr> mCounterNoiseMap[ncrates][4];      /// Counter for the Noise Hit Map
```
Comment (njacazio) [inline Modules/TOF/include/TOF/TaskRaw.h:85]: If you feel particularly majestic you could add in the comment that we have one per crate and one per FEA (the documentation will pay off in the future!)

File: Modules/TOF/include/TOF/TaskRaw.h [hunk 2]
```diff
@@ -158,6 +159,7 @@ class TaskRaw final : public TaskInterface
   std::shared_ptr<TH1F> mHistoIndexEO;          /// Index in electronic
   std::shared_ptr<TH1F> mHistoIndexEOInTimeWin; /// Index in electronic for noise analysis
   std::shared_ptr<TH1F> mHistoIndexEOIsNoise;   /// Noise hit map x channel
+  std::shared_ptr<TH1F> mHistoRDHTriggers;      /// RDH served and received trigger efficiency
```
Comment (njacazio) [inline Modules/TOF/include/TOF/TaskRaw.h:162]: Does this have also the trigger efficiency? The comment is not so clear I fear

File: Modules/TOF/src/TaskRaw.cxx [hunk 3]
```diff
@@ -47,9 +47,23 @@ void RawDataDecoder::rdhHandler(const o2::header::RAWDataHeader* rdh)
 {
   mCounterRDH[rdh->feeId & 0xFF].Count(0);
 
+  //Fatal
```
Comment (njacazio) [inline Modules/TOF/src/TaskRaw.cxx:50]: Maybe here it is worth to extend a bit the comment, fatal is a bit too harsh maybe and we can consider printing warning messages, what do you think?

File: Modules/TOF/src/TaskRaw.cxx [hunk 4]
```diff
@@ -47,9 +47,23 @@ void RawDataDecoder::rdhHandler(const o2::header::RAWDataHeader* rdh)
...
   if ((rdh->detectorField & 0x00010000) != 0) {
     mCounterRDH[rdh->feeId & 0xFF].Count(1);
   }
+
+  //Trigger expected and served
+  if (rdh->stop) { // if RDH close
+    int triggerserved = ((rdh->detectorField & 0xFF00070F) >> 24);
```
Comment (njacazio) [inline Modules/TOF/src/TaskRaw.cxx:57]: let's `const` these bad boys

File: Modules/TOF/src/TaskRaw.cxx [hunk 5]
```diff
@@ -47,9 +47,23 @@ void RawDataDecoder::rdhHandler(const o2::header::RAWDataHeader* rdh)
...
+    int triggerreceived = ((rdh->detectorField & 0x00FF070F) >> 16);
+    if (triggerserved < triggerreceived) {
+      //Trigger error
+      mCounterRDH[rdh->feeId & 0xFF].Count(2);
+    }
+    //Trigger efficiency
```
Comment (njacazio) [inline Modules/TOF/src/TaskRaw.cxx:63]: This is not strictly the trigger efficiency, maybe you can consider adding: `Ingredients for the trigger efficiency` (also for the future)

File: Modules/TOF/src/TaskRaw.cxx [hunk 6]
```diff
@@ -432,8 +449,10 @@ void TaskRaw::endOfCycle()
       mHistoSlotParticipating->SetBinContent(crate + 1, j + 4, mDecoderRaw.mCounterTRM[crate][j].HowMany(0));
     }
     mHistoSlotParticipating->SetBinContent(crate + 1, 1, mDecoderRaw.mCounterRDH[crate].HowMany(0));
+    if (mDecoderRaw.mCounterRDHTriggers[1].HowMany(crate) != 0) {
+      mHistoRDHTriggers->SetBinContent(crate + 1, (double)mDecoderRaw.mCounterRDHTriggers[0].HowMany(crate) / mDecoderRaw.mCounterRDHTriggers[1].HowMany(crate));
```
Comment (njacazio) [inline Modules/TOF/src/TaskRaw.cxx:453]: maybe consider using the explicitly the cast i.e. `static_cast<double>(`... `)` or `float`

**Comment by njacazio** [comment]
Ciao @ercolessi great work as usual. I have few very minor comments to improve the code but other than that outstanding job!

**Comment by njacazio** [comment]
@Barthelemy @knopers8 now that we have more people working on the TOF QC project could we have some codeowners for the TOF directory so that we are triggered when something is changed ? Is this possible?

File: Modules/TOF/src/TaskRaw.cxx (hunk 3 above)
Comment (ercolessi) [reply Modules/TOF/src/TaskRaw.cxx:50]: Ciao @njacazio, many thanks for the comments!
So, fatal is how Roberto refers to it in https://github.com/AliceO2Group/AliceO2/pull/5802/commits/90c8e48b4701eb30e5c0aa90acc517e983e112c3 .
How would you suggest to change it?

For the warning: sounds great, where you thinking of a LOG(WARNING) or using a checker?

Comment (njacazio) [reply Modules/TOF/src/TaskRaw.cxx:50]: I mean, I don't think we want to throw a fatal error if we encounter this word here, therefore I would at least call it `Case for the RDH word: fatal` or something that to explain that it is not an error in our task.
I guess you can put the LOG and we will implement the checker as well to inform the QC crew.
Does it seem reasonable?

Comment (ercolessi) [reply Modules/TOF/src/TaskRaw.cxx:50]: Everything clear, sounds perfect! I'm on it

File: Modules/TOF/src/TaskRaw.cxx [hunk 7]
```diff
@@ -491,11 +511,12 @@ void TaskRaw::reset()
   mHistoIndexEOInTimeWin->Reset();
   mHistoTimeBC->Reset();
   mHistoIndexEOIsNoise->Reset();
+  mHistoRDHTriggers->Reset();
 
   mDecoderRaw.resetHistograms();
 }
 
-const char* RawDataDecoder::RDHDiagnosticsName[2] = { "RDH_HAS_DATA", "RDH_DECODER_FATAL" };
+const char* RawDataDecoder::RDHDiagnosticsName[3] = { "RDH_HAS_DATA", "RDH_DECODER_FATAL", "RDH_TRIGGER_ERROR" };
```
Comment (njacazio) [inline Modules/TOF/src/TaskRaw.cxx:519]: Maybe you could use the new `nRDHwords`

**Review by njacazio — APPROVED** [review]
Ciao Francesca, for me it's top quality. Few suggestions but I would go forth.
Thanks a lot!
Nicolo'

File: Modules/TOF/include/TOF/TaskRaw.h [hunk 8]
```diff
@@ -61,19 +61,20 @@ class RawDataDecoder final : public DecoderBase
...
-  static const char* DRMDiagnosticName[nwords]; /// DRM Counter names
-  static const char* LTMDiagnosticName[nwords]; /// LTM Counter names
-  static const char* TRMDiagnosticName[nwords]; /// TRM Counter names
+  static const char* RDHDiagnosticsName[nRDHwords]; /// RDH Counter names
+  static const char* DRMDiagnosticName[nwords];     /// DRM Counter names
+  static const char* LTMDiagnosticName[nwords];     /// LTM Counter names
+  static const char* TRMDiagnosticName[nwords];     /// TRM Counter names
   // Diagnostic counters
-  Counter<2, RDHDiagnosticsName> mCounterRDH[ncrates];            /// RDH Counters
+  Counter<3, RDHDiagnosticsName> mCounterRDH[ncrates];            /// RDH Counters
```
[the remark this answers is not in the document]
Comment (ercolessi) [reply Modules/TOF/include/TOF/TaskRaw.h:77]: yes indeed thanks!

**Comment by njacazio** [comment]
@Barthelemy @knopers8 I think we are done now and are ready to merge as soon as the tests are passed.
Btw, now that we have more people working on the TOF QC project could we have some codeowners for the TOF directory so that we are triggered when something is changed ? Would that be possible?

**Comment by Barthelemy** [comment]
@njacazio I am going to take a quick look and approve it. Concerning CODEOWNERS I think that it is a good idea. I'll look into it (no idea how it works).

**Comment by Barthelemy** [comment]
https://alice.its.cern.ch/jira/browse/QC-605

File: Modules/TOF/src/TaskRaw.cxx [hunk 9]
```diff
@@ -47,8 +47,23 @@ void RawDataDecoder::rdhHandler(const o2::header::RAWDataHeader* rdh)
 {
   mCounterRDH[rdh->feeId & 0xFF].Count(0);
 
+  //Case for the RDH word "fatal"
   if ((rdh->detectorField & 0x00010000) != 0) {
     mCounterRDH[rdh->feeId & 0xFF].Count(1);
+    LOG(WARNING) << "RDH flag \"fatal\" error occurred in crate " << static_cast<int>(rdh->feeId & 0xFF);
```
Comment (Barthelemy) [inline Modules/TOF/src/TaskRaw.cxx:53]: InfoLogger is used in some places and FairLogger in some others. It is ok for me, just raising your awareness about it.

**Review by Barthelemy — APPROVED** [review]

File: Modules/TOF/src/TaskRaw.cxx (hunk 9 above)
Comment (njacazio) [reply Modules/TOF/src/TaskRaw.cxx:53]: Thanks! We will address it in the next PR

**Pull request merged**